### DISCUSSION TITLE
Using Swarm in CI/CD - how to get status of update?

### DISCUSSION THREAD

**Discussion by ivanduka** [BretFisher/ama#137] [Q&A]
Hi Bret, I am going through your course and it's amazing!
I have a question about using Swarm in CI/CD context.

Currently, my workflow is the following:
1. I commit and push my code to GitHub
2. GitHub actions kick in and create a Docker image of my app, uploads it to the Docker Hub
3. GitHub actions login via SSH to the target server and run `docker service update $SERVICE_NAME --image ivanduka/$APP_NAME:$APP_VERSION`

This setup works amazing due to healthchecks and automatic load balancing, I practically have zero downtime deployment for free. At the same time, if something is wrong with my containers, Swarm does not route the traffic to them and does the rollback.

However, one thing still puzzles me: how do I check if the `docker service update ...` goes wrong?
For example, I tried to deploy a version of my app with a healthcheck that always fails.
Swarm tries to deploy it, does the retries and then reports that the update failed and does the rollback.
Basically, the docker service command never fails.
As a result, my GitHub Actions pipeline reports success and I am oblivious of this failed deployment.

What I want is to have a way to know that the image update failed. I googled for a couple of days and cannot find the answer, which puzzles me - Swarm must be used in thousands of organizations to deploy apps. How do they check if it succeeds or fails in CI/CD context?

**Comment by BretFisher** [comment]
Assuming you're not using `-d` in your update commands, then it will wait for success or failure. If it fails, docker returns a exit code of 1, which Actions should be picking up as a failed step in its workflow.  That would be the way you knew it didn't deploy correctly.

**Comment by ivanduka** [comment]
Unfortunately, this works only for `docker service update` as `docker stack deploy` is always executed in asynchronous manner.

**Comment by BretFisher** [comment]
Correct. `stack deploy` is also a CLI-only feature, so there's no API to work against, making it hard to automate.
